src/Components/Analyzers/src/InternalUsageAnalyzer.cs

```diff
@@ -35,87 +35,148 @@ public InternalUsageAnalyzer(Func<ISymbol, bool> isInInternalNamespace, Func<ISy
-                            {
-                                context.ReportDiagnostic(Diagnostic.Create(_descriptor, creationSyntax.Type.GetLocation(), containingType));
-                                return;
-                            }
-                        }
+                case IFieldSymbol field:
+                    VisitDeclarationSymbol(context, field.Type, field);
+                    break;
 
-                        return;
+                case IMethodSymbol method:
+
+                    // Ignore return types on property-getters. Those will be reported through
+                    // the property analysis.
+                    if (method.MethodKind != MethodKind.PropertyGet)
+                    {
+                        VisitDeclarationSymbol(context, method.ReturnType, method);
                     }
 
-                case ClassDeclarationSyntax declarationSyntax:
+                    // Ignore parameters on property-setters. Those will be reported through
+                    // the property analysis.
+                    if (method.MethodKind != MethodKind.PropertySet)
                     {
-                        if (context.SemanticModel.GetDeclaredSymbol(declarationSyntax)?.BaseType is ISymbol symbol &&
-                            symbol.ContainingAssembly != context.Compilation.Assembly &&
-                            (IsInInternalNamespace(symbol) || HasInternalAttribute(symbol)) &&
-                            declarationSyntax.BaseList?.Types.Count > 0)
+                        foreach (var parameter in method.Parameters)
                         {
-                            context.ReportDiagnostic(Diagnostic.Create(_descriptor, declarationSyntax.BaseList.Types[0].GetLocation(), symbol));
+                            VisitDeclarationSymbol(context, parameter.Type, method);
                         }
-
-                        return;
                     }
+                    break;
 
-                case ParameterSyntax parameterSyntax:
-                    {
-                        if (context.SemanticModel.GetDeclaredSymbol(parameterSyntax)?.Type is ISymbol symbol &&
-                            symbol.ContainingAssembly != context.Compilation.Assembly &&
-                            (IsInInternalNamespace(symbol) || HasInternalAttribute(symbol)))
-                        {
+                case IPropertySymbol property:
+                    VisitDeclarationSymbol(context, property.Type, property);
+                    break;
 
-                            context.ReportDiagnostic(Diagnostic.Create(_descriptor, parameterSyntax.GetLocation(), symbol));
-                        }
+                case IEventSymbol @event:
+                    VisitDeclarationSymbol(context, @event.Type, @event);
+                    break;
+            }
+        }
 
-                        return;
-                    }
+        // Similar logic here to VisitDeclarationSymbol, keep these in sync.
```

Comment:
I think you can probably do couple of tweaks here:

1. Extract out the logic into a common helper `void VisitSymbol(ISymbol symbol, Action<Diagnostic> reportDiagnostic, Location location, SymbolDisplayFormat format)`
2. Maintain a per-compilation cache of processed symbols so you avoid duplicate processing of symbols and also duplicate diagnostics for the same symbol.

Reply:
> Maintain a per-compilation cache of processed symbols so you avoid duplicate processing of symbols and also duplicate diagnostics for the same symbol.

How would this help?
- We want to report diagnostics on *usage* of symbols defined in our library (operations)
- We want to report diagnostics on *declarations* based on symbols defined in our library (symbols)

My assumption is that our symbol action is going to be called once-per-declaration in user code - preventing duplicates in reporting errors for those sites.

Reply:
> Extract out the logic into a common helper void VisitSymbol(ISymbol symbol, Action<Diagnostic> reportDiagnostic, Location location, SymbolDisplayFormat format)

I wanted to avoid creating a delegate per operation/symbol (`reportDiagnostic). There's ways around it but dey nasty. 

```C#
        private void VisitSymbol<TContext>(TContext context, Compilation compilation, Action<TContext, Location> reportDiagnostic, ISymbol symbol, Location location)
        {
            if (symbol.ContainingAssembly == compilation.Assembly)
            {
                // This is part of the compilation, avoid this analyzer when building from source.
                return;
            }

            if (HasInternalAttribute(symbol))
            {
                reportDiagnostic(context, location);
                return;
            }

            var containingType = symbol.ContainingType;
            if (IsInInternalNamespace(containingType) || HasInternalAttribute(containingType))
            {
                reportDiagnostic(context, location);
                return;
            }
        }
```

I think I'd prefer the code duplication.